server/finatra-server/src/main/scala/tapir/server/finatra/package.scala

```diff
@@ -0,0 +1,102 @@
+package tapir.server
+import java.nio.charset.Charset
+
+import com.twitter.finagle.http.{Request, Response, Status}
+import com.twitter.inject.Logging
+import com.twitter.util.Future
+import tapir.DecodeResult.{Error, Mismatch, Missing, Multiple}
+import tapir.EndpointInput.PathCapture
+import tapir.internal.server.{DecodeInputs, DecodeInputsResult, InputValues}
+import tapir.internal.{SeqToParams, _}
+import tapir.{DecodeFailure, DecodeResult, Endpoint, EndpointIO, EndpointInput}
+
+import scala.reflect.ClassTag
+import scala.util.control.NonFatal
+
+package object finatra {
+  implicit class RichFinatraEndpoint[I, E, O](e: Endpoint[I, E, O, Nothing]) extends Logging {
+    def toRoute(logic: I => Future[Either[E, O]]): FinatraRoute = {
+
+      val handler = { request: Request =>
+        def decodeBody(result: DecodeInputsResult): DecodeInputsResult = {
+          result match {
+            case values: DecodeInputsResult.Values =>
+              values.bodyInput match {
+                case Some(bodyInput @ EndpointIO.Body(codec, _)) =>
+                  val rawBody =
+                    FinatraRequestToRawBody(codec.meta.rawValueType, request.content, request.charset.map(Charset.forName), request)
+
+                  codec.safeDecode(Some(rawBody)) match {
+                    case DecodeResult.Value(bodyV) => values.value(bodyInput, bodyV)
+                    case failure: DecodeFailure    => DecodeInputsResult.Failure(bodyInput, failure): DecodeInputsResult
+                  }
+                case None => values
+              }
+            case failure: DecodeInputsResult.Failure => failure
+          }
+        }
+
+        def valuesToResponse(values: DecodeInputsResult.Values): Future[Response] = {
+          val i = SeqToParams(InputValues(e.input, values.values)).asInstanceOf[I]
+          logic(i)
+            .map {
+              case Right(result) => OutputToFinatraResponse(e.output, result).toResponse
+              case Left(err)     => OutputToFinatraResponse(e.errorOutput, err, None, Status.BadRequest).toResponse
+            }
+            .onFailure {
+              case NonFatal(ex) =>
+                error(ex)
+            }
+        }
+
+        def handleDecodeFailure(
```

Comment:
why not use the `DecodeFailureHandler` mechanism?

Reply:
This was a simplification which avoided the `ServerOptions` -- but now that you bring it up I realize that I missed this and should do a `ServerOptions` so that we can pass in a `DecodeFailureHandler` -- so I'll add a `ServerOptions` and give it a decodeFailureHandler field.

Reply:
👍